gax/src/test/java/com/google/api/gax/batching/v2/FlowControllerTest.java

```
@@ -0,0 +1,253 @@
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(2)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+    flowController.reserve(1);
+
+    testBlockingReserveRelease(flowController, 5);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByNumberOfBytes() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingElementCount(100)
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    testBlockingReserveRelease(flowController, 10);
+  }
+
+  @Test
+  public void testReserveRelease_blockedByNumberOfBytes_noElementCountLimit() throws Exception {
+    FlowController flowController =
+        new FlowController(
+            FlowControlSettings.newBuilder()
+                .setMaxOutstandingRequestBytes(10L)
+                .setLimitExceededBehavior(LimitExceededBehavior.Block)
+                .build());
+
+    testBlockingReserveRelease(flowController, 10);
+  }
+
+  private static void testBlockingReserveRelease(
+      final FlowController flowController, final int maxNumBytes) throws Exception {
+
+    flowController.reserve(1);
+
+    final SettableFuture<?> permitsReserved = SettableFuture.create();
+    Future<?> finished =
+        Executors.newCachedThreadPool()
+            .submit(
+                new Runnable() {
+                  @Override
+                  public void run() {
+                    try {
+                      permitsReserved.set(null);
+                      flowController.reserve(maxNumBytes);
+                    } catch (FlowControlException e) {
+                      throw new AssertionError(e);
```

I think you should just let the exception bubble up on its own